plugins/module_utils/common.py

```diff
@@ -668,14 +669,18 @@ def _empty_resource_list():
             else:
                 # Delete the object
                 result['changed'] = True
-                if not self.check_mode:
+                if self.check_mode and not self.supports_dry_run:
```

Comment:
Instead of doing these checks on two values, can we instead wrap `check_mode` in a setter (e.g: `enable_check_mode()`), and check the `supports_dry_run value` inside the setter. This way, check_mode can be true only if supports_dry_run is true. And we do the check at one single place in the code.

Reply:
I may not be fully understanding what you are suggesting, but I don't think this would be possible. The problem is check_mode can be true even if the library doesn't support dry run. In that case, we use the older, slightly buggy client side dry run. So there are really three states a module could be in: 1. no check mode 2. check mode with dry run 3. check mode without dry run. Different actions will be taken for all three states.

I agree this is ugly. It's a perfect example of one of the problems I've highlighted in this collection as why it needs refactoring. This is a change that ideally happens in one place. Until we can drop support for the older kubernetes client I think it's always going to be a bit messy, but once we can solely rely on dry run the best way to handle this is to probably be able to put the client into something like read only mode. This would happen once during module setup and we wouldn't need to have any check mode logic in the rest of the code.

Reply:
Thanks for taking the time to explain the problem.